### PR TITLE
use builtin Fractions to help with Rational and igcd

### LINKED ISSUE #11081: where possible, use python fractions for Rational
Since we have dropped support for 2.5 we can use the [fractions module](https://docs.python.org/2/library/fractions.html?highlight=rational) for Rational internals. The `gcd` method could also replace `igcd`:

```python
>>> timeit('gcd(12,3)','from fractions import gcd')
0.5296611342553904
>>> timeit('gcd(12,3)','from sympy import igcd as gcd')
6.766568780548827
```

### PR DESCRIPTION
**Pull request by smichr**
closes #11081 
closes #11089
- `igcd` now uses fractions.gcd to do the computation; caching is still in place. Also, a check that _all_ args be ints will catch cases where a gcd=1 might occur before bad input is encountered.
- the `Fraction` object is used to do handle some parsing of `Rational` input and to compute the limited denominator. I wonder if it might be better to use `Fraction` as the underlying representation and then update methods like `Rational.p` and `Rational.args` to return the correct args of the `Fraction`. Any thoughts would be welcome.
- [x] clean up history

### CODE REVIEW

**Comment by smichr** [comment]
ping @isuruf -- please take a look at the mods here in contrast to #11089

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -155,24 +155,17 @@ def igcd(*args):
...
-                    b = -b
-
-                while b:
-                    a, b = b, a % b
-            else:
-                a = abs(a or b)
-
-            _gcdcache[(a, b)] = a
-        if a == 1 or b == 1:
+        a = _gcdcache.setdefault((a, b), gcd(a, b))
```
Comment (isuruf) [inline sympy/core/numbers.py]: `gcd(a, b)` is computed each time `igcd(a, b)` is called.
```
try:
    a = _gcdcache[(a, b)]
except KeyError:
    a = _gcdcache[(a, b)] = gcd(a, b)
```

**Comment by isuruf** [comment]
> I wonder if it might be better to use Fraction as the underlying representation and then update methods like Rational.p and Rational.args to return the correct args of the Fraction

Implementation of Fractions doesn't do the `gcd` hint trick introduced in #11019, therefore will be slower to use. Also, arithmetic operators would use `gcd` instead of `igcd` and therefore caching will not be used in that case.

File: sympy/core/numbers.py [hunk 2]
```diff
@@ -155,24 +155,17 @@ def igcd(*args):
     5
 
     """
+    from fractions import gcd
```
Comment (isuruf) [inline sympy/core/numbers.py]: You can also import the gcd function from `gmpy` if `HAS_GMPY` is true

**Comment by smichr** [comment]
ping @isuruf - ok, have a look now. I benchmark with

```
for I in range(10**4):
 for J in range(10**4):
   g = igcd(I,j)
```

I think it is now more similar to the original `igcd` now.

File: sympy/core/numbers.py [hunk 3]
```diff
@@ -155,26 +155,32 @@ def igcd(*args):
...
+        a = args.pop()
+        if a == 0:
+            continue
+        break
+    a = abs(as_int(a))
+    while args:
+        b = args.pop()
+        if b == 0:
+            continue
+        b = as_int(b)
```
Comment (isuruf) [inline sympy/core/numbers.py]: Why not have abs here?

File: sympy/core/numbers.py [hunk 4]
```diff
@@ -155,26 +155,32 @@ def igcd(*args):
...
-            return 1
-    return a
+            while b:
+                a, b = b, a % b
+            if a < 0:
+                a = -a
+            if a == 1:
+                test_others = [as_int(a) for a in args]
+                return 1
+            _gcdcache[t] = a
```
Comment (isuruf) [inline sympy/core/numbers.py]: This should come before line 179

Comment (smichr) [inline sympy/core/numbers.py]: I was thinking not to cache the trivial arg but `a` is now the result. You're right. I'll move it.

File: sympy/core/numbers.py (hunk 3 above)
Comment (smichr) [inline sympy/core/numbers.py]: Although the sign change is handled below, handling it here would keep the cache from including negatives. I'll change it.

**Comment by smichr** [comment]
I feel like the rewritten `igcd` is sound and ready for review, @isuruf and @jksuom . The code I used for testing is in the commit message.

File: sympy/core/numbers.py [hunk 5]
```diff
@@ -155,25 +155,30 @@ def igcd(*args):
...
+    a = abs(as_int(args[0])) if args else 0
+    k = 1
+    if a != 1:
+        while k < len(args):
+            b = args[k]
+            k += 1
+            try:
+                a = _gcdcache[(a, b)]
+            except KeyError:
+                b = as_int(b)
```
Comment (jksuom) [inline sympy/core/numbers.py]: Should this be done before looking into the cache?

File: sympy/core/numbers.py [hunk 6]
```diff
@@ -155,25 +155,30 @@ def igcd(*args):
...
+                t = a, b
                 while b:
                     a, b = b, a % b
-            else:
-                a = abs(a or b)
-
-            _gcdcache[(a, b)] = a
-        if a == 1 or b == 1:
-            return 1
+                _gcdcache[t] = _gcdcache[t[1], t[0]] = a
```
Comment (jksuom) [inline sympy/core/numbers.py]: Should `t[1], t[0]` be a tuple?

File: sympy/core/numbers.py (hunk 5 above)
Comment (smichr) [inline sympy/core/numbers.py]: My thought was this: S(1) in {1:2} is True so the type doesn't matter. The usual use case is to use `igcd` on int/Integers so we can get the answer before doing a function call to `as_int`.

File: sympy/core/numbers.py (hunk 6 above)
Comment (smichr) [inline sympy/core/numbers.py]: It behaves as a tuple:

```
>>> d={};d[1,2]=3;d
{(1, 2): 3}
```

Comment (jksuom) [inline sympy/core/numbers.py]: I thought so. (It was only that there are parentheses above on line 185.)

Comment (isuruf) [inline sympy/core/numbers.py]: Instead of `_gcdcache[t] = _gcdcache[t[1], t[0]] = a`, you can simply do `_gcdcache[t] = a`

**Comment by jksuom** [comment]
This looks good to me. +1

File: sympy/core/numbers.py (hunk 6 above)
Comment (smichr) [inline sympy/core/numbers.py]: By this line I am entering (a,b) and (b,a) into the cache.

Comment (isuruf) [inline sympy/core/numbers.py]: +1

**Pull request merged**